theseus/geometry/so3.py

```diff
@@ -0,0 +1,231 @@
+        )
+        theta_nz = torch.where(small_theta, non_zero, theta)
+        theta2_nz = torch.where(small_theta, non_zero, theta2)
+        a = torch.where(small_theta, 8 / (4 + theta2) - 1, theta.cos())
+        b = torch.where(small_theta, 0.5 * a + 0.5, theta.sin() / theta_nz)
+        c = torch.where(small_theta, 0.5 * b, (1 - a) / theta2_nz)
+        ret.data = c * tangent_vector.view(-1, 3, 1) @ tangent_vector.view(-1, 1, 3)
+        ret[:, 0, 0] += a.view(-1)
+        ret[:, 1, 1] += a.view(-1)
+        ret[:, 2, 2] += a.view(-1)
+        temp = b.view(-1, 1) * tangent_vector
+        ret[:, 0, 1] -= temp[:, 2]
+        ret[:, 1, 0] += temp[:, 2]
+        ret[:, 0, 2] += temp[:, 1]
+        ret[:, 2, 0] -= temp[:, 1]
+        ret[:, 1, 2] -= temp[:, 0]
+        ret[:, 2, 1] += temp[:, 0]
+        return ret
+
+    def _log_map_impl(self) -> torch.Tensor:
+        ret = torch.zeros(self.shape[0], 3, dtype=self.dtype, device=self.device)
+        ret[:, 0] = 0.5 * (self[:, 2, 1] - self[:, 1, 2])
+        ret[:, 1] = 0.5 * (self[:, 0, 2] - self[:, 2, 0])
+        ret[:, 2] = 0.5 * (self[:, 1, 0] - self[:, 0, 1])
+        cth = 0.5 * (self[:, 0, 0] + self[:, 1, 1] + self[:, 2, 2] - 1)
+        sth = ret.norm(dim=1)
+        theta = torch.atan2(sth, cth)
+        # theta != pi
+        not_near_pi = 1 + cth > 1e-7
+        # Compute the approximation of theta / sin(theta) when theta is near to 0
+        small_theta = theta[not_near_pi] < 5e-3
+        non_zero = torch.ones(1, dtype=self.dtype, device=self.device)
+        sth_nz = torch.where(small_theta, non_zero, sth[not_near_pi])
+        scale = torch.where(
+            small_theta, 1 + sth[not_near_pi] ** 2 / 6, theta[not_near_pi] / sth_nz
+        )
+        ret[not_near_pi] *= scale.view(-1, 1)
+        # theta ~ pi
+        near_pi = ~not_near_pi
+        ddiag = torch.diagonal(self[near_pi], dim1=1, dim2=2)
+        # Find the index of major coloumns and diagonals
+        major = torch.logical_and(
+            ddiag[:, 1] > ddiag[:, 0], ddiag[:, 1] > ddiag[:, 2]
+        ) + 2 * torch.logical_and(ddiag[:, 2] > ddiag[:, 0], ddiag[:, 2] > ddiag[:, 1])
+        ret[near_pi] = self[near_pi, major]
+        ret[near_pi, major] -= cth[near_pi]
+        ret[near_pi] *= (theta[near_pi] ** 2 / (1 - cth[near_pi])).view(-1, 1)
+        ret[near_pi] /= ret[near_pi, major].sqrt().view(-1, 1)
+        return ret
+
+    def _compose_impl(self, so3_2: LieGroup) -> "SO3":
+        raise NotImplementedError
+
+    def _inverse_impl(self, get_jacobian: bool = False) -> "SO3":
+        return SO3(data=self.data.transpose(1, 2).clone())
+
+    def to_matrix(self) -> torch.Tensor:
+        return self.data.clone()
+
+    def to_quaternion(self) -> torch.Tensor:
+        raise NotImplementedError
+
+    @staticmethod
+    def hat(tangent_vector: torch.Tensor) -> torch.Tensor:
+        _check = tangent_vector.ndim == 3 and tangent_vector.shape[1:] == (3, 1)
```

Comment:
There are tangent vector checks in other functions that could be consolidated out into a single function.